### ISSUE TITLE
Next.js 8 and RNW Transpilation Issues

### ISSUE THREAD

**Issue by dcalhoun** [zeit/next.js#7276]
# Bug report

## Describe the bug

**TL;DR: I need to transpile ESM source with Next.js, but I am unable to get it working with Next.js 8**

I am attempting to transpile `react-native-web` source inside of my Next.js project. I succeeded at doing this in the past ([see older Next.js RNW example](https://github.com/zeit/next.js/tree/ba5f1b2c212d254463640c7a6405d74d777c9215/examples/with-react-native-web)). With Next.js 8, transpiling is no longer working. Note, the current [Next.js + RNW example]() circumvents this issue by [not transpiling at all](https://github.com/zeit/next.js/commit/dbdd66392158d3890fd89818fdd52b32bd399b73) (i.e. it relies on the CJS modules, not the ESM).

I have a [test case](https://github.com/dcalhoun/nextjs-rnw-transpile) that includes individual commits for attempted [usage of custom webpack config](https://github.com/dcalhoun/nextjs-rnw-transpile/tree/fe1b1ba48e6515061d9892a9e0a9c510542314a3), [usage of `next-transpile-modules`](https://github.com/dcalhoun/nextjs-rnw-transpile/tree/1b3ebc4a3f0c9dead48de65a6f4277d6623f1e53), and [usage of babel config only](https://github.com/dcalhoun/nextjs-rnw-transpile/tree/7079d1ceb407207a6bedc2f20562b79110e58023).

For the first two approaches (webpack and `next-transpile-modules`), the compilation fails with the following error:

<details>
<summary>Click to view error details</summary>

```
TypeError: (0 , _compile.classic) is not a function
    at /Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:8788:45
    at Array.forEach (<anonymous>)
    at Object.createCSS (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:8786:33)
    at Object.create (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:8899:35)
    at Object../node_modules/react-native-web/dist/cjs/modules/createDOMProps/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:17329:28)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Object../node_modules/react-native-web/dist/cjs/exports/createElement/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:14409:46)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Object../node_modules/react-native-web/dist/cjs/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:14708:45)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Module../pages/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:32163:70)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Object.3 (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:32206:18)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at /Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:91:18
    at Object.<anonymous> (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:94:10)
TypeError: (0 , _compile.classic) is not a function
    at /Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:8788:45
    at Array.forEach (<anonymous>)
    at Object.createCSS (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:8786:33)
    at Object.create (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:8899:35)
    at Object../node_modules/react-native-web/dist/cjs/modules/createDOMProps/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:17329:28)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Object../node_modules/react-native-web/dist/cjs/exports/createElement/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:14409:46)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Object../node_modules/react-native-web/dist/cjs/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:14708:45)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Module../pages/index.js (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:32163:70)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at Object.3 (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:32206:18)
    at __webpack_require__ (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:23:31)
    at /Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:91:18
    at Object.<anonymous> (/Users/davidcalhoun/Sites/oss/nextjs-rnw-transpile/.next/server/static/development/pages/index.js:94:10)
```

</details>

For the babel-only attempt, compilation succeeds, but it is loading the CJS modules, not the ESM (i.e. it is not transpiling and merely using the already transpiled RNW source, [RNW provides both](https://github.com/necolas/react-native-web/blob/801937748b2f3c96284bee1881164ac0c62a9c6d/packages/react-native-web/package.json#L8-L9)).

At this point, I have been unable to determine what has changed in Next.js that is causing these issues. So, my questions are:

1. What changed in Next.js that would impact webpack config/`next-transpile-modules`?
2. I noticed [changes](https://github.com/zeit/next.js/commit/0f23faf81feca8e8526161448d64c7a5d665a38e#diff-b90c160ab7612f85353c0d6d25eac21cR32) to [webpack `externals`](https://github.com/zeit/next.js/commit/a5ad06f99b597615e91e3fc8e96dd11e6f5f6f5f), how does that potentially impact my goal?

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Clone and run the [test case](https://github.com/dcalhoun/nextjs-rnw-transpile).
1. Visit the app server in the browser.

## Expected behavior
The Next.js app successfully compiles `react-native-web` source.

## System information
 - OS: macOS 10.14.4
 - Browser: n/a
 - Version of Next.js: 8.1.0

## Additional context

This was originally reported in the [Spectrum community](https://spectrum.chat/next-js/general/next-js-8-and-rnw-transpilation-issues~d414072e-1590-4bde-9a32-0cbdedf0c199). I posted here as well due to lack of feedback and the potential that it is in fact a bug with Next.

**Comment by Aleksion** [comment]
@dcalhoun ant update on this? I have a fully working NextJS 9 repo with RNW etc. But I'm pretty sure it includes ALL of RNW meaning the bundle is... massive

**Comment by dcalhoun** [comment]
@aleksion unfortunately, no, I have not made progress on this. My app is currently frozen on Next.js 7.x for now.

**Comment by flybayer** [comment]
@dcalhoun @Aleksion I'm able to compile and run your test RNW project by adding [babel-plugin-react-native-web](https://github.com/necolas/react-native-web/tree/master/packages/babel-plugin-react-native-web).

I opened a PR with the changes: https://github.com/dcalhoun/nextjs-rnw-transpile/pull/1/files

However, I'm not sure if the end result is exactly what you're looking for, but it definitely compiles and runs.

**Comment by flybayer** [comment]
I made a PR to add `babel-plugin-react-native-web` to solve another issue, but like I said in my previous comment I think it solves this issue too.

https://github.com/zeit/next.js/pull/9078

Please confirm.

**Comment by dcalhoun** [comment]
@flybayer thanks for the information. I don't have time to confirm the results currently, but I will report back my findings eventually. Hopefully I have time to review later this week.

**Comment by dcalhoun** [comment]
Thanks to @flybayer, I was able to resolve my issue. I don't fully understand what is happening deep down in the compilation, but including [`babel-plugin-react-native-web`](https://github.com/necolas/react-native-web/tree/master/packages/babel-plugin-react-native-web) in the project's `babel.config.js` does resolve this issue.

I did notice what difference between [the approach in this PR](https://github.com/dcalhoun/nextjs-rnw-transpile) vs [the current Next.js RNW example](https://github.com/zeit/next.js/tree/v9.1.2-canary.7/examples/with-react-native-web) project. The former uses ESM for both the server and client. The latter uses CJS for the server and ESM for the client. I'm not sure if there is a downside to relying on CJS for the server, it may be fine. For my specific project, I at least need RNW's ESM for the client.

Thanks again to @flybayer for the help!

**Comment by timneutkens** [comment]
I don't believe we should close this if it's not clear what is causing the exact issue and the solution doesn't seem correct to me still.

**Comment by dcalhoun** [comment]
@timneutkens fair enough. I looked again at this and have [discovered that upgraded Next to the latest (e.g. `9.x.x`) has now addressed my original issue](https://github.com/dcalhoun/nextjs-rnw-transpile/commit/8e790138ebfdf21674eabf253fbc49164fea236c) (i.e. the addition of `babel-plugin-react-native-web` is not required). I have not researched heavily enough to know where the fix was introduced in Next's version history, but my original issue is now resolved without requiring changes. Something between `8.x.x` and `9.x.x` changed and resolved this.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
